Refactor: Move CUDA graph dispatch logic earlier

yiz-liu opened the pull request:
## Purpose
After carefully reading the code, I found a potential edge case: when `execute_dummy_batch` runs, dummy attention metadata isn't created even if `aclgraph_runtime_mode` is later set to `CUDAGraphMode.FULL`. That's odd, because attention normally requires metadata, otherwise it may raise an error or produce incorrect output.

The only explanation I can think of is that we're skipping metadata creation for dummy batches to save a bit of performance since we don't care about their output. Can anyone elaborate on this? Thanks.

I also propose a potential fix by moving CUDA graph dispatch logic earlier, this ensures metadata is built when replaying a CUDA graph, and the performance impact should be negligible.

## Test Plan
None.

## Test Result
None.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

fhl2000 commented:
Yep, it is a good catch. The flashInfer backend may potentially hang if it is at full cudagraph without preparing attn_metadata. Could you please move the code of attn_metadata building back out of the `with` context, instead, moving the cudagraph dispatching parts above?

yiz-liu commented:
> Yep, it is a good catch. The flashInfer backend may potentially hang if it is at full cudagraph without preparing attn_metadata. Could you please move the code of attn_metadata building back out of the `with` context, instead, moving the cudagraph dispatching parts above?

OK, will do.

yiz-liu committed:
Refactor: Move CUDA graph dispatch logic earlier

Moves the CUDA graph dispatch logic to execute before the attention metadata is calculated within the dummy run.

Signed-off-by: Yizhou Liu <liu_yizhou@outlook.com>

fhl2000 reviewed (approved):
Sorry for this late review. I think there is no harm in moving this logic earlier in dummy_run.
cc @LucasWilkinson, it is also closer to your idea in that the padding logic (and cg mode) verified by the cudagraph dispatcher is done before attention metadata building.

fhl2000 commented:
cc @ProExpertProg

ProExpertProg reviewed (approved):
LGTM, @LucasWilkinson is this ok with you?

LucasWilkinson reviewed (approved):
LGTM; will be fixed by: https://github.com/vllm-project/vllm/pull/28579 but we can take this in the interim

yiz-liu committed:
Merge branch 'main' into main
